### PR TITLE
KAFKA-17273: Migrate BootstrapControllersIntegrationTest to use ClusterTestExtensions

### PR DESCRIPTION
**Pull request by FrankYang0529**
The other sub tasks need to define custom configs for broker/controller, and hence using `ClusterTestExtensions` is more suitable to this test class.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/test/java/kafka/test/server/BootstrapControllersIntegrationTest.java [hunk 1]
```diff
@@ -142,64 +103,71 @@ public void testPutControllersInBootstrapBrokersConfig() {
...
-    public void testDescribeCluster(boolean usingBootstrapControllers) throws Exception {
-        try (Admin admin = Admin.create(adminProperties(usingBootstrapControllers))) {
-            DescribeClusterResult result = admin.describeCluster();
-            assertEquals(cluster.controllers().values().iterator().next().clusterId(),
-                    result.clusterId().get(1, TimeUnit.MINUTES));
-            if (usingBootstrapControllers) {
-                assertEquals(((QuorumController) cluster.waitForActiveController()).nodeId(),
-                    result.controller().get().id());
+    @ClusterTest
+    public void testDescribeCluster(ClusterInstance clusterInstance) {
```
Comment (chia7712) [inline core/src/test/java/kafka/test/server/BootstrapControllersIntegrationTest.java]: Could you please split it into 2 test cases? for example:
```java
    @ClusterTest
    public void testDescribeClusterByControllers(ClusterInstance clusterInstance) throws Exception {
        testDescribeCluster(clusterInstance, true);
    }

    @ClusterTest
    public void testDescribeCluster(ClusterInstance clusterInstance) throws Exception {
        testDescribeCluster(clusterInstance, false);
    }

    private void testDescribeCluster(ClusterInstance clusterInstance, boolean usingBootstrapControllers) throws Exception {
        try (Admin admin = Admin.create(adminConfig(clusterInstance, usingBootstrapControllers))) {
            DescribeClusterResult result = admin.describeCluster();
            assertEquals(clusterInstance.clusterId(), result.clusterId().get(1, TimeUnit.MINUTES));
            if (usingBootstrapControllers) {
                assertTrue(clusterInstance.controllerIds().contains(result.controller().get().id()));
            }
        }
    }
```

File: core/src/test/java/kafka/test/server/BootstrapControllersIntegrationTest.java [hunk 2]
```diff
@@ -142,64 +103,71 @@ public void testPutControllersInBootstrapBrokersConfig() {
...
-            DescribeFeaturesResult result = admin.describeFeatures();
-            short metadataVersion = cluster.controllers().values().iterator().next().
-                featuresPublisher().features().metadataVersion().featureLevel();
-            assertEquals(new FinalizedVersionRange(metadataVersion, metadataVersion),
-                result.featureMetadata().get(1, TimeUnit.MINUTES).finalizedFeatures().
-                    get(MetadataVersion.FEATURE_NAME));
-        }
+    @ClusterTest
+    public void testDescribeFeatures(ClusterInstance clusterInstance) {
+        Stream.of(false, true).forEach(usingBootstrapControllers -> {
```
Comment (chia7712) [inline core/src/test/java/kafka/test/server/BootstrapControllersIntegrationTest.java]: ditto

File: core/src/test/java/kafka/test/server/BootstrapControllersIntegrationTest.java [hunk 3]
```diff
@@ -142,64 +103,71 @@ public void testPutControllersInBootstrapBrokersConfig() {
...
-                    () -> result.all().get(1, TimeUnit.MINUTES));
-            assertNotNull(exception.getCause());
-            assertEquals(InvalidUpdateVersionException.class, exception.getCause().getClass());
-            assertTrue(exception.getCause().getMessage().endsWith("does not support this feature."),
-                "expected message to end with 'does not support this feature', but it was: " +
-                    exception.getCause().getMessage());
-        }
+    @ClusterTest
+    public void testUpdateFeatures(ClusterInstance clusterInstance) {
+        Stream.of(false, true).forEach(usingBootstrapControllers -> {
```
Comment (chia7712) [inline core/src/test/java/kafka/test/server/BootstrapControllersIntegrationTest.java]: ditto

File: core/src/test/java/kafka/test/server/BootstrapControllersIntegrationTest.java [hunk 4]
```diff
@@ -142,64 +103,71 @@ public void testPutControllersInBootstrapBrokersConfig() {
...
-    @ValueSource(booleans = {false, true})
-    public void testDescribeMetadataQuorum(boolean usingBootstrapControllers) throws Exception {
-        try (Admin admin = Admin.create(adminProperties(usingBootstrapControllers))) {
-            DescribeMetadataQuorumResult result = admin.describeMetadataQuorum();
-            assertEquals(((QuorumController) cluster.waitForActiveController()).nodeId(),
-                result.quorumInfo().get(1, TimeUnit.MINUTES).leaderId());
-        }
+    @ClusterTest
+    public void testDescribeMetadataQuorum(ClusterInstance clusterInstance) throws Exception {
+        Stream.of(false, true).forEach(usingBootstrapControllers -> {
```
Comment (chia7712) [inline core/src/test/java/kafka/test/server/BootstrapControllersIntegrationTest.java]: ditto

File: core/src/test/java/kafka/test/server/BootstrapControllersIntegrationTest.java [hunk 5]
```diff
@@ -212,33 +180,36 @@ public void testUsingBootstrapControllersOnUnsupportedAdminApi() {
...
-                    filter(e -> e.name().equals("my.custom.config")).
-                    findFirst().get();
-                assertEquals(DYNAMIC_BROKER_CONFIG, entry.source(),
-                    "Expected entry for my.custom.config to come from DYNAMIC_BROKER_CONFIG. " +
-                    "Instead, the entry was: " + entry);
-            });
-        }
+    @ClusterTest
+    public void testIncrementalAlterConfigs(ClusterInstance clusterInstance) {
+        Stream.of(false, true).forEach(usingBootstrapControllers -> {
```
Comment (chia7712) [inline core/src/test/java/kafka/test/server/BootstrapControllersIntegrationTest.java]: ditto

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 thanks for this nice migration!

**Comment by chia7712** [comment]
@FrankYang0529 could you please rebase code to trigger QA again

**Comment by FrankYang0529** [comment]
> @FrankYang0529 could you please rebase code to trigger QA again

Yes, rebased it. Thank you.

**Comment by chia7712** [comment]
The failed tests are traced by following jira

1. https://issues.apache.org/jira/browse/KAFKA-13514
2. https://issues.apache.org/jira/browse/KAFKA-17312
3. https://issues.apache.org/jira/browse/KAFKA-14396

**Review by chia7712 — APPROVED** [review]
LGTM

**Pull request merged**